File: c++/0646.cpp
// Link: https://leetcode.com/problems/maximum-length-of-pair-chain/description/
// Runtime: 66 ms
// Memory: 26.02 MB

class Solution 
{
public:
    int findLongestChain(const std::vector<std::vector<int>>& pairs) 
    {
        std::vector<std::vector<int>> sortedPairs = pairs;
        std::sort(sortedPairs.begin(), sortedPairs.end(), [](const auto& a, const auto& b) {return a[1] < b[1];});

        int lengthOfChain = 1, currentRightBorder = sortedPairs[0][1];
        for (int i = 1; i < static_cast<int>(sortedPairs.size()); i++)
            if (currentRightBorder < sortedPairs[i][0])
            {
                lengthOfChain++;
                currentRightBorder = sortedPairs[i][1];
            }
        
        return lengthOfChain;
    }
};

File: python/0646.py
# Link: https://leetcode.com/problems/maximum-length-of-pair-chain/description/
# Runtime: 189 ms
# Memory: 16.96 MB


class Solution:
    def findLongestChain(self, pairs: List[List[int]]) -> int:
        sorted_pairs = pairs[:]
        sorted_pairs.sort(key=lambda pair: pair[1])

        length_of_chain, current_right_border = 1, sorted_pairs[0][1]
        for i in range(1, len(sorted_pairs)):
            if current_right_border < sorted_pairs[i][0]:
                length_of_chain += 1
                current_right_border = sorted_pairs[i][1]

        return length_of_chain
